File: bindings/node/src/zlink/contracts/sockets/stream_socket.ts
// SPDX-License-Identifier: MPL-2.0

import type { RoutingId } from '../core';
import type { Received, SendOperation, StreamPacket } from '../messaging';
import type { RecvFlags } from './socket_constants';
import type { StreamSocketOptions } from './socket_options';
import type { Socket } from './socket';

/**
 * STREAM socket: exchanges framed packets with raw TCP peers addressed by
 * routing id.
 */
export interface StreamSocket extends Socket {
  /** The STREAM-specific typed options facade. */
  readonly options: StreamSocketOptions;
  /**
   * Begin an exact-target managed send. `submit()` resolves only after Core
   * accepts the record for the routing id captured by this builder.
   */
  send(routingId: RoutingId): SendOperation;
  /** Receive a message into `result`; false when `RecvFlags.DontWait` is set and none is available. */
  recv(result: Received, flags?: RecvFlags): boolean;
  /** Receive one packet into reusable storage; false for non-blocking no-data. */
  recvPacket(result: StreamPacket, flags?: RecvFlags): boolean;
  /**
   * Set the routing id that identifies this socket to its peers. Apply before
   * connecting so peers observe it from the first packet.
   */
  setRoutingId(routingId: RoutingId): void;
  /** Return the routing id that identifies this socket to its peers. */
  getRoutingId(): RoutingId;
  /** Disconnect the peer identified by `routingId`. */
  disconnectRid(routingId: RoutingId): void;
}

File: bindings/python/src/zlink/contracts/sockets/stream_socket.py
# SPDX-License-Identifier: MPL-2.0

from typing import Protocol, runtime_checkable

from ..core.routing_id import RoutingId
from ..messaging.received import Received
from ..messaging.stream_packet import StreamPacket
from . import socket as _socket_contract
from .codes import RecvFlags
from .operations import SendOp


@runtime_checkable
class StreamSocket(_socket_contract._SocketContract, Protocol):
    """STREAM socket: exchanges framed packets with raw TCP peers."""

    @property
    def stream_options(self):
        """The STREAM-specific typed options facade."""
        ...

    def send(self, routing_id: RoutingId) -> SendOp:
        """Begin a send addressed to ``routing_id``."""
        ...

    def recv_into(
        self, received: Received, *, flags: RecvFlags = RecvFlags.NONE
    ) -> bool:
        """Receive a message into ``received`` storage; ``False`` when
        ``DONT_WAIT`` is set and none is available."""
        ...

    def recv_packet_into(
        self, out: StreamPacket, *, flags: RecvFlags = RecvFlags.NONE
    ) -> bool:
        """Receive one framed packet into reusable caller-owned storage."""
        ...

    def disconnect_rid(self, peer_rid):
        """Disconnect the peer identified by ``peer_rid``."""
        ...


__all__ = ["StreamSocket"]
